gdctools/gdc_mirror.py

```diff
@@ -182,14 +206,44 @@ def __mirror_file(self, file_d, proj_root, n, total, retries=3):
             if retry > retries:
                 # A partially downloaded file will interfere with subsequent
                 # mirrors
-                common.silent_rm(savepath)
-                logging.error("Error downloading file {0}, too many retries ({1})".format(savepath, retries))
+                common.silent_rm(partialsavepath)
+                logging.warning("Error downloading file {0}, too many retries ({1})".format(savepath, retries))
             else:
-                #Save md5 checksum on success
+                #success! First remove old md5, then rename file, then write new md5
+                #That way, if the process gets interrupted, future runs will be consider
+                # the file dirty unless each of these steps complete.
                 md5sum = file_d['md5sum']
                 md5path = savepath + ".md5"
+                if os.path.exists(md5path):
+                    os.remove(md5path)
+                os.rename(partialsavepath,savepath)
                 with open(md5path, 'w') as mf:
                     mf.write(md5sum + "  " + basename)
+                file_status = 'pass'
+
+        return file_status
+
+    @staticmethod
+    def __merge_mirror_metadata(old, new):
+        pass
+        file_item_dict = {}
+        for file_item in old:
+            key = file_item['file_id'] + file_item['md5sum']
+            file_item_dict[key] = file_item
+
```

Comment:
this really feels like it should be a list comprehension
```python
file_item_dict = {(file_item['file_id'] + file_item['md5sum']): file_item for file_item in old}
```

Reply:
Could be.  I prefer this idiom.